powerauth-backend-tests/src/test/java/com/wultra/security/powerauth/test/v3/PowerAuthSignatureTest.java

```diff
@@ -573,6 +575,92 @@ void signatureOfflinePersonalizedValidTest() throws Exception {
+
+        // The remainder of last line is Base64 encoded ECDSA signature
+        final String ecdsaSignature = lastLine.substring(1);
+        final byte[] serverPublicKeyBytes = Base64.getDecoder().decode((String) model.getResultStatusObject().get("serverPublicKey"));
+        final ECPublicKey serverPublicKey = (ECPublicKey) config.getKeyConvertor().convertBytesToPublicKey(serverPublicKeyBytes);
+
+        // Prepare offline data without signature
+        final String offlineDataWithoutSignature = offlineDataResponse.substring(0, offlineDataResponse.length() - ecdsaSignature.length());
+
+        // Validate ECDSA signature of data using server public key
+        assertTrue(signatureUtils.validateECDSASignature(offlineDataWithoutSignature.getBytes(StandardCharsets.UTF_8), Base64.getDecoder().decode(ecdsaSignature), serverPublicKey));
+
+        // Prepare data for PowerAuth signature
+        final String proximityTotp = expectedResult ? parts[5] : "11111111";
+        final String dataForSignature = operationId + "&" + operationData + "&" + proximityTotp;
+
+        // Prepare normalized data for signature
+        final String signatureBaseString = PowerAuthHttpBody.getSignatureBaseString("POST", "/operation/authorize/offline", Base64.getDecoder().decode(nonce), dataForSignature.getBytes(StandardCharsets.UTF_8));
+
+        // Prepare keys
+        byte[] signaturePossessionKeyBytes = Base64.getDecoder().decode((String) model.getResultStatusObject().get("signaturePossessionKey"));
+        byte[] signatureKnowledgeKeySalt = Base64.getDecoder().decode((String) model.getResultStatusObject().get("signatureKnowledgeKeySalt"));
+        byte[] signatureKnowledgeKeyEncryptedBytes = Base64.getDecoder().decode((String) model.getResultStatusObject().get("signatureKnowledgeKeyEncrypted"));
+
+        // Get the signature keys
+        final SecretKey signaturePossessionKey = config.getKeyConvertor().convertBytesToSharedSecretKey(signaturePossessionKeyBytes);
+        final SecretKey signatureKnowledgeKey = EncryptedStorageUtil.getSignatureKnowledgeKey(config.getPassword().toCharArray(), signatureKnowledgeKeyEncryptedBytes, signatureKnowledgeKeySalt, new KeyGenerator());
+
+        // Put keys into a list
+        final List<SecretKey> signatureKeys = new ArrayList<>();
+        signatureKeys.add(signaturePossessionKey);
+        signatureKeys.add(signatureKnowledgeKey);
+
+        // Calculate signature of normalized signature base string with 'offline' as application secret
+        final String signature = signatureUtils.computePowerAuthSignature((signatureBaseString + "&offline").getBytes(StandardCharsets.UTF_8), signatureKeys, CounterUtil.getCtrData(model, stepLogger), SignatureConfiguration.decimal());
+
+        final VerifyOfflineSignatureRequest verifyRequest = new VerifyOfflineSignatureRequest();
+        verifyRequest.setActivationId(config.getActivationIdV3());
+        verifyRequest.setData(signatureBaseString);
```

Comment:
This is completely wrong. Here we have to put data without OTP because it is unknown to the callee.